crates/rune-macros/src/context.rs

```diff
@@ -358,6 +378,67 @@ impl Context {
                             }
                         },
                     });
+                } else if meta.path == GET_REF {
+                    attr.field = true;
+                    attr.protocols.push(FieldProtocol {
+                        custom: self.parse_field_custom(meta.input)?,
+                        generate: |g| {
+                            let Generate {
+                                ty,
+                                target,
+                                ..
+                            } = g;
+
+                            match target {
+                                GenerateTarget::Named { field_ident, field_name } => {
+                                    let protocol = g.tokens.protocol(PROTOCOL_GET);
+
+                                    match ty {
+                                        syn::Type::Reference(TypeReference{ mutability: None, ..}) => {
+                                            quote_spanned! { g.field.span() =>
+                                                module.field_function(#protocol, #field_name, |s: &Self| {
+                                                    let s = Shared::new(AnyObj::new_projection(s.#field_ident));
+                                                    Value::Any(s)
+                                                })?;
+                                            }
+                                        },
+                                        syn::Type::Reference(TypeReference{ mutability: Some(_), ..}) => {
+                                            quote_spanned! { g.field.span() =>
+                                                module.field_function(#protocol, #field_name, |s: &mut Self| {
```

Comment:
This also looks a bit sketchy to me. Since the return `Shared<T>` contains a mutable projection from `&mut Self`, we need to ensure that the same projection cannot legally be performed multiple time.

Consider what would happen if we mutably reference the same field may times.

Reply:
That all makes sense! I'm struggling to get it working, though.

We can't map a `Mut<T>` unless it's owned. There's a blanket implementation preventing us from using anything other than an `Any` to get an owned value for the field function:
```rust
impl<T> FromValue for Mut<T>
where
    T: Any,
{
    fn from_value(value: Value) -> VmResult<Self> {
        VmResult::Ok(vm_try!(vm_try!(value.into_any()).downcast_into_mut()))
    }
}
```
One option is to make a second kind of `Mut<T>` that works specifically with `AnyRef`. But, for `FromValue` to be satisfied, we'd still need to drop the `'static` constraint.

Alternatively, there is `UnsafeToRef` and `UnsafeToMut`, but I don't think we're able to map from those.